biz.aQute.bndlib/src/aQute/bnd/osgi/Clazz.java

```diff
@@ -1550,19 +1565,28 @@ private void doTypeAnnotations(DataInput in, ElementType member, RetentionPolicy
 
 			//
 			// Rest is identical to the normal annotations
-			doAnnotation(in, member, policy, false, access_flags);
+			Annotation annotation = doAnnotation(in, member, policy, target_index, collect, access_flags);
+			if (cd != null && annotation != null) {
+				cd.annotation(annotation);
+			}
 		}
 	}
 
 	private void doAnnotations(DataInput in, ElementType member, RetentionPolicy policy, int access_flags)
```

Comment:
And instead of using -1 everywhere, please specify a constant in `Annotation` like 

`public static final int TARGET_INDEX_NONE = -1;`

 and use that constant. Then it is self documenting.